tests/cxx/typedef_in_template.cc

```diff
@@ -24,34 +25,87 @@ class Container {
 
 
 void Declarations() {
-  // These do not need the full type for Class because they're template params.
+  // Just using Container does not need the full type for Class because there
+  // are only aliases made, which do not require full-uses.
 
-  // TODO: This is almost certainly wrong, see bug #431
-  // We should not require the full definition of Class for passing it as a
-  // template argument, but we must require it when the typedef it's aliasing
-  // is full-used.
-  // The bug has instructions for how to provoke the error more obviously.
+  // TODO: But currently this is counted as a full-use because Class is used
+  // inside a template specialization (of Pair) within the definition of
+  // Container.
+  // IWYU: Class is...*typedef_in_template-i1.h
+  // IWYU: Class needs a declaration
+  Container<Class> c;
+
+  // Full-using any of those aliases *should* require a full use of Class.
 
+  // IWYU: Class is...*typedef_in_template-i1.h
   // IWYU: Class needs a declaration
   Container<Class>::value_type vt;
 
+  // IWYU: Class is...*typedef_in_template-i1.h
   // IWYU: Class needs a declaration
   Container<Class>::pair_type pt;
 
+  // IWYU: Class is...*typedef_in_template-i1.h
   // IWYU: Class needs a declaration
   Container<Class>::alias_type at;
 }
 
+// STL containers are often implemented via a complex web of type aliases and
+// helper classes.  Tracking uses through all these layers can be non-trivial.
+// The following are some reduced examples in roughly increasing order of
+// complexity which can serve as helpful test cases while debugging such
+// issues.  They were inspired by libstdc++'s implementation of
+// std::unordered_map, but don't directly correspond to it.
+
+// Verify that a full-use of an alias of a template parameter is treated as a
+// full-use of that parameter.
+template <typename T>
+struct UsesAliasedParameter {
+  using TAlias = T;
+  TAlias t;
+};
+
+// IWYU: IndirectClass is...*indirect.h
+// IWYU: IndirectClass needs a declaration
```

Comment:
At first I found it suspicious that there was both a full- and a forward-decl-use here, but I think it kind-of makes sense. The forward-decl use is for the `using TAlias = T;`, and the full use is for the `TAlias t;`?

I guess I could argue that the forward-decl use does not belong here, because it's an internal detail of the template, but it doesn't really do any harm either, since it will be short-circuited by the full use.

Reply:
After having seen a number of other template testcases during review, I think I now understand that the forward-decl is for the use of the template argument. Nothing to see here.